gui: add metrics for txn chart (#4603)
firedancer-io/firedancer: book/api/websocket.md

| transactions         | `number\|null` | Total number of transactions (vote and non-vote) in the block. If the slot is not skipped, this will be non-null, but in some cases it will also be non-null even if the slot was skipped. That's because we replayed the block but selected a fork without it, but we still know how many transactions were in it |
| vote_transactions    | `number\|null` | Total number of vote transactions in the block. Will always be less than or equal to `transactions`. The number of non-vote transactions is given by `transactions - vote_transactions`
| failed_transactions  | `number\|null` | Total number of failed transactions (vote and non-vote) in the block. Failed transactions are those which are included in the block and were charged fees, but failed to execute successfully. This is different from dropped transactions which do not pay fees and are not included in the block |

> | max_compute_units                 | `number`    | The maximum number of compute units that can be packed into the slot.  This limit is one of many consensus-critical limits defined by the solana protocol, and helps keeps blocks small enough for validators consume them quickly.  It may grow occasionally via on-chain feature activations |

Reviewer: nit: alignment

> | max_compute_units                 | `number`    | The maximum number of compute units that can be packed into the slot.  This limit is one of many consensus-critical limits defined by the solana protocol, and helps keeps blocks small enough for validators consume them quickly.  It may grow occasionally via on-chain feature activations |

Reviewer: Extreme nit but it might be good to rename these to `cost_units` to align with the new RPC field. Up to you though, 
just consider it, maybe it's less clear that way.

| waterfall           | `TxnWaterfall\|null`      | If the slot is not `mine`, will be `null`. Otherwise, a waterfall showing reasons transactions were acquired since the end of the prior leader slot |
| tile_primary_metric | `TilePrimaryMetric\|null` | If the slot is not `mine`, will be `null`. Otherwise, max value of per-tile-type primary metrics since the end of the prior leader slot |
| tile_timers         | `TsTileTimers[]\|null`    | If the slot is not `mine`, will be `null`. Otherwise, an array of `TsTileTimers` samples from the slot, sorted earliest to latest. We store this information for the most recently completed 4096 leader slots. This will be `null` for leader slots before that |

> | transactions        | `Transactions\|null`      | If the slot is not `mine`, will be `null`. Otherwise, metrics for the transactions in this slot. Arrays have a seperate entry for each transaction that was packed in this slot, and are ordered in the same order the transactions were packed. |

Reviewer: Extreme style nits for operator facing documentation: no period at end of these docs
```suggestion
| transactions        | `Transactions\|null`      | If the slot is not `mine`, will be `null`. Otherwise, metrics for the transactions in this slot. Arrays have a seperate entry for each transaction that was packed in this slot, and are ordered in the same order the transactions appear in the block |
```

| timestamp_nanos   | `string`      | A timestamp of when the tile timers were sampled, nanoseconds since the UNIX epoch |
| tile_timers       | `TileTimer[]` | A list of all tile timing information at the given sample timestamp |

**`Transactions`**
| Field                             | Type        | Description |
|-----------------------------------|-------------|-------------|
| max_compute_units                 | `number`    | The maximum number of compute units that can be packed into the slot |
| start_timestamp_nanos             | `string`    | A UNIX timestamp in nanoseconds, representing the time that we started packing transactions into the slot |

> | target_end_timestamp_nanos        | `string`    | A UNIX timestamp in nanoseconds, representing the target time in nanoeconds that we should stop packing transactions for the slot. Transactions might still finish executing after this end time, if they started executing before it and ran over the deadline |

Reviewer: ```suggestion
| target_end_timestamp_nanos        | `string`    | A UNIX timestamp in nanoseconds, representing the target time in nanoeconds that we should stop packing transactions for the slot. Transactions might still finish executing after this end time, if they started executing before it and ran over the deadline. In rare cases, transactions may also appear to begin after this timestamp due to slight clock drift between execution cores |
```

> | txn_start_timestamps_nanos        | `string[]`  | An array of UNIX timestamps, in nanoseconds. `txn_start_timestamps_nanos[i]` is approximatley the time when the microblock for the `i`-th transaction in the slot began executing. |

Reviewer: Why is this approximate, and not exact?

Author: Since there is some time between this timestamp (when pack constructs constructs the frag and sends it off to bank) and when the microblock actually begins to execute in the recieved bank, though I guess it really depends on what "start executing" means here. Ill change it to something more exact/descriptive.

| txn_stop_timestamps_nanos         | `string[]`  | An array of UNIX timestamps, in nanoseconds. `txn_stop_timestamps_nanos[i]` is approximatley the time when the microblock for the `i`-th transaction in the slot completed executing. |
| txn_compute_units_requested       | `number[]`  | `txn_compute_units_requested[i]` is the requested (via computeBudgetInstruction) or default compute units for the `i`-th transaction in the slot. |
| txn_compute_units_estimated       | `number[]`  | `txn_compute_units_estimated[i]` is the estimated (by pack) cost for the `i`-th transaction in the slot, as determined by its payload contents and compute budget. |

> | txn_compute_units_rebated         | `number[]`  | `txn_compute_units_rebated[i]` is the difference between the cost estimate and the actual cost (which is strictly less than or equal to the estimate) for the `i`-th transaction in the slot. |

Reviewer: I would replace this with `compute_units_consumed`, and the developer can derive the rebate if they want. Most developers are never going to care about the rebate, but reading "consumed" is very clear.

| txn_priority_fee                  | `string[]`  | `txn_priority_fee[i]` is the priority fee in lamports for the `i`-th transaction in the slot. |

> | txn_tips                          | `string[]`  | `txn_tips[i]` is the total tip in lamports for the `i`-th transaction in the slot. Non-bundle transactions have a tip of 0. |

Reviewer: If I was building an app with this field, I wouldn't understand what it means without looking at the code. Does this include the Jito commission? Is it just the part going to the validator, or does it include the staker tip portion also? Is it a specific tip program / account (Jito), or consolidated across what might be multiple?

> | txn_error_code                    | `number[]`  | `txn_error_code[i]` is the error code that explains the failure for the `i`-th transaction in the slot. If the error code is 0, then the transaction succeeded. |

Reviewer: No period after description

InvalidRentPayingAccount                | 28   | Deprecated. |
WouldExceedMaxVoteCostLimit             | 29   | Exceeded the maximum vote compute unit cost allowed for this slot. |
WouldExceedAccountDataTotalLimit        | 30   | Deprecated. |
DuplicateInstruction                    | 31   | Contains duplicate instructions. |
InsufficientFundsForRent                | 32   | Deprecated. |
MaxLoadedAccountsDataSizeExceeded       | 33   | Retreived accounts data size exceeds the limit imposed for this transaction. |
InvalidLoadedAccountsDataSizeLimit      | 34   | Requested an invalid data size (i.e. 0). |
ResanitizationNeeded                    | 35   | Sanitized transaction differed before/after feature activiation. Needs to be resanitized. |
ProgramExecutionTemporarilyRestricted   | 36   | Execution of a program referenced by this transaciton is restricted. |
UnbalancedTransaction                   | 37   | The total accounts balance before the transaction does not equal the total balance after. |
ProgramCacheHitMaxLimit                 | 38   | The program cache allocated for transaction batch for this transaction hit its load limit. |

> CommitCancelled                         | 39   | This transaction was part of a bundle that failed. |

Reviewer: Is this `39` correct? I think we have our own bundle failed error code (`BundlePeer`)

Author: I'm using the agave error codes here. This one is based on [this](https://github.com/firedancer-io/agave/blob/bcec2ca2e0218ca5470a2a5360ca520dc44898a7/core/src/banking_stage/committer.rs#L146)

| txn_compute_units_requested       | `number[]`  | `txn_compute_units_requested[i]` is the requested (or protocol-defined default if no computeBudget instruction is provided) execute compute units for the `i`-th transaction in the slot. This amount does not include other transaction-derived costs, such as signature costs, account data costs, or trasaction data costs |
| txn_max_compute_units             | `number[]`  | `txn_compute_units_estimated[i]` is a strict upper bound on the total cost for the `i`-th transaction in the slot.  The transaction cannot have succeeded if its incurred cost (known after execution) exceeds this bound.  This bound is used by the pack tile to estimate the pace at which the block is being filled, and to filter out transactions that it knows will fail ahead of time. |

> | txn_compute_units_consumed        | `number[]`  | `txn_compute_units_consumed[i]` is the actual post-execution cost of `i`-th transaction in the slot.  While some transactions costs are known from the transaction payload itself (such as the cost incurred by the amount of instruction data), other costs (like execution costs or the cost due to loaded on-chain account data) are a function of the state of the blockchain at the time of execution. This value represents the actual cost after a transaction is executed.  Consensus requires that all validators agree on this value for a given transaction in a slot.  This value will be 0 for transactions in a failed bundle, since these transactions are never included in the block.  Regular failed transactions may still consume compute units if they are included in the block |

Reviewer: > This value will be 0 for transactions in a failed bundle
I'm not sure on this API design, because you've hid information. Now there's no way to see consumed CUs of a bundle txn that failed. If you just showed it, the caller can not count it if it sees it's a failed bundle.

Author: good point, will change

> | txn_priority_fee                  | `string[]`  | `txn_priority_fee[i]` is the priority fee in lamports for the `i`-th transaction in the slot.  The priority fee is computed by multiplying the requested execution cost (derived from a provided computeBudget instruction, or from a protocol defined default) by the compute unit price (derived from a seperate computeBudget instruction).  Note that this array includes the priority fee parsed from failed transactions as well, even though those transactions do not pay this fee |

Reviewer: Failed transactions do pay the priority fee?

Author: I probably meant transactions not included in the block (e.g. failed txns from a bundle). Will change/clarify.

| txn_exec_end_timstamps_nanos      | `string[]`  | An array of UNIX timestamps, in nanoseconds. `txn_exec_end_timstamps_nanos[i]` is the time when the `i`-th transaction in the slot finished executing in the SVM. At this point, the execution result is known and the transaction is ready to be committed to the block. |
| txn_microblock_id                 | `string[]`  | `txn_microblock_id[i]` is the index of the microblock for the `i`-th transaction in the slot.  Microblocks are collections of 1+ transactions.  All of the transactions from a bundle share the same microblock. Microblock ids are monotonically increasing in the order they appear in the block and start at 0 for each slot |

These are the possible error codes that might be included in `txn_error_code` and their meanings.

| Code Name                             | Code | Description |
|---------------------------------------|------|-------------|

> Success                                 | 0    | The transaction successfully executed |

Reviewer: I think the table formatting should be
```suggestion
| Success                                 | 0    | The transaction successfully executed |
```
